### PR TITLE
python 3.11 runtime

### PR DESCRIPTION
**Pull request by jefimm**

### CODE REVIEW

File: gaepypi/_decorators.py [hunk 1]
```diff
@@ -14,59 +14,49 @@
...
-
-                else:
-                    return func(handler, *args, **kwargs)
+def load_accounts():
+	ret = {}
+	with open('config.json') as data_file:
+		config = json.load(data_file)
+	for account in config["accounts"]:
+		ret[account['username']] = account
+	return ret
```
Comment (yaelmi3) [inline gaepypi/_decorators.py:29]:
```suggestion
	with open('config.json') as data_file:
		config = json.load(data_file)
	return {account['username']: account for account in config["accounts"]}
```

Comment (yaelmi3) [reply gaepypi/_decorators.py:29]: also I prefer using pathlib for opening files, but it's mostly cosmetic

File: gaepypi/_decorators.py [hunk 2]
```diff
@@ -14,59 +14,49 @@
...
-    handler.response.set_status(403, message="Forbidden")
 
+def valid_credentials(username, password, required_roles=None):
+	account = account_by_name.get(username)
+	if not account or account["password"] != __basic_hash(password):
+		return False
+	user_roles = account.get('roles', [])
+	if required_roles and any([(required_role not in user_roles) for required_role in required_roles]):
+		return False
+	return True
```
Comment (yaelmi3) [inline gaepypi/_decorators.py:48]:
```suggestion
	return not required_roles or all(required_role in user_roles for required_role in required_roles)
```

File: gaepypi/_decorators.py [hunk 3]
```diff
@@ -14,59 +14,49 @@
...
 
-def __basic_lookup(username):
-    with open('config.json') as data_file:
-        config = json.load(data_file)
-    for account in config["accounts"]:
-        if account['username'] == username:
-            return account
 
+def basic_auth(required_roles=None):
+	def inner_decorator(f):
```
Comment (yaelmi3) [inline gaepypi/_decorators.py:52]: use `@functools.wraps(f)` and then you don't need this: `wrapped.__name__ = f.__name__`

File: gaepypi/_decorators.py [hunk 4]
```diff
@@ -14,59 +14,49 @@
...
-        return callf
-    return decorator_basic_auth
 
+account_by_name = load_accounts()
 
-def __basic_login(handler):
-    handler.response.set_status(401, message="Authorization Required")
-    handler.response.headers['WWW-Authenticate'] = 'Basic realm="Secure Area"'
 
+def __basic_hash(password):
```
Comment (yaelmi3) [inline gaepypi/_decorators.py:35]: why dunder and not just private?

File: gaepypi/package.py [hunk 5]
```diff
@@ -79,15 +79,13 @@ def to_html(self):
...
-    @contextmanager
     def get_file(self, filename, storage=None):
         if filename not in self.files:
             raise GAEPyPIError("File not found for {0}".format(self))
         storage = self.enquire_storage(storage)
         path = storage.get_package_path(self.name, self.version, filename)
         gcs_file = storage.read(path)
-        yield gcs_file
-        gcs_file.close()
+        return gcs_file
```
Comment (yaelmi3) [inline gaepypi/package.py:88]:
```suggestion
        return storage.read(path)
```

File: gaepypi/storage.py [hunk 6]
```diff
@@ -20,13 +20,16 @@
...
-                                          backoff_factor=2,
-                                          max_retry_period=15)
-gcs.set_default_retry_params(my_default_retry_params)
+storage_client = storage.Client()
+
+
+def to_bucket_and_path(p):
+	segments = p.lstrip('/').split('/')
+	bucket_path = '/'.join(segments[1:])
+	return segments[0], bucket_path
```
Comment (yaelmi3) [inline gaepypi/storage.py:32]: `segments[0]` never seem to be used it'a always `_, bucket_path = to_bucket_and_path(path)`, so why do we return it?

Comment (yaelmi3) [inline gaepypi/storage.py:32]:
```suggestion
    _, *segments = path.lstrip('/').split('/')
    bucket_path = '/'.join(segments)
    return bucket_path
```

File: gaepypi/storage.py [hunk 7]
```diff
@@ -138,22 +141,48 @@ def split_path(self, path):
...
+
+        ret = []
+        for blob in blobs:
+            ret.append(self._legacy_path(blob.name))
+
+        if dir_only:
+            ret = []
+            for p in blobs.prefixes:
+                ret.append(self._legacy_path(p))
+        return ret
```
Comment (yaelmi3) [inline gaepypi/storage.py:156]: padded = path if path[-1] == '/' else f'{path}/'
        _, bucket_path = to_bucket_and_path(padded)
        blobs = storage_client.list_blobs(self.bucket, prefix=bucket_path, delimiter='/')

        ret = [self._legacy_path(blob.name) for blob in blobs]
        if dir_only:
            ret = [self._legacy_path(p) for p in blobs.prefixes]
        return ret

Comment (yaelmi3) [reply gaepypi/storage.py:156]: also checkout https://pypi.org/project/pathy/, [example](https://github.com/CyCognito/mayhem/blob/0a2899aae563696b563b2a0d393b9d41f641fd91/mayhem/gcloud_storage.py#L25-L28) - you might like it :)

File: gaepypi/storage.py [hunk 8]
```diff
@@ -138,22 +141,48 @@ def split_path(self, path):
...
+
+    def _legacy_path(self, p):
+        return '/' + self.bucket + '/' + p
```
Comment (yaelmi3) [inline gaepypi/storage.py:159]:
```suggestion
        return f'/{self.bucket}/{p}'
```

File: gaepypi/storage.py [hunk 9]
```diff
@@ -138,22 +141,48 @@ def split_path(self, path):
...
+        return blob.exists()
 
     def path_exists(self, path):
-        match = list(gcs.listbucket(path.rstrip('/'), delimiter='/'))
-        return path.rstrip('/') in [stat.filename.rstrip('/') for stat in match]
+        _, bucket_path = to_bucket_and_path(path)
+        blobs = storage_client.list_blobs(self.bucket, prefix=bucket_path, delimiter='/')
+        ret = {}
+        for blob in blobs:
+            ret.add(blob.name)
```
Comment (yaelmi3) [inline gaepypi/storage.py:187]: I assume you need `ret` to be a `set`, but using `{}` considered a dict.
so you need `ret = set()`
also:
```
ret = {blob.name for blob in blobs}
``` is more pythonic

File: main.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+app.wsgi_app = wrap_wsgi_app(app.wsgi_app)
+storage = None
+
+
+def get_storage():
+	global storage
+	if not storage:
+		bucket_name = os.environ.get('BUCKET_NAME')
+		if not bucket_name:
+			bucket_name = app_identity.get_default_gcs_bucket_name()
```
Comment (yaelmi3) [inline main.py]:
```suggestion
		bucket_name = os.environ.get('BUCKET_NAME') or app_identity.get_default_gcs_bucket_name()
```

File: main.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+@app.route("/", methods=['POST'])
+@basic_auth(required_roles=['write'])
+def root_post():
+	name = request.values.get('name')
+	version = request.values.get('version')
+	action = request.values.get(':action')
+
+	if name and version and len(request.files) and action == 'file_upload':
+		try:
```
Comment (yaelmi3) [inline main.py:43]: I always prefer to put try/except closer to the potential issue, so in this case I'd put it before `package = Package(get_storage(), name, version)`

File: main.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+			package.put_file(filename, upload.stream)
+		except GAEPyPIError as e:
+			abort(403)
+	return ""
+
+
+@app.route("/packages", methods=['GET'])
+@basic_auth()
+def packages_get():
+	storage = get_storage()
```
Comment (yaelmi3) [inline main.py]: Shadows name 'storage' from outer scope, is it by design?

File: main.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+		return package.to_html()
+	abort(404)
+
+
+@app.route("/packages/<name>/<version>/<filename>", methods=['GET'])
+@basic_auth()
+def package_download(name, version, filename):
+	try:
+		package = Package(get_storage(), name, version)
+		gcs_file = package.get_file(filename)
```
Comment (yaelmi3) [inline main.py:87]: I think is what should be wrapped in try/except

File: tests/unit/test_gcstorage.py [hunk 14]
```diff
@@ -33,89 +33,89 @@ def test_split_path(self):
...
-    def test_write(self, retry, open):
-        retry.return_value = 'rp'
-        m = Mock()
-        m.write = Mock()
-        open.return_value = m
-        self.s.write('/mybucket/path/a.txt', '1111111')
-        retry.assert_called_with(backoff_factor=1.1)
-        open.assert_called_with('/mybucket/path/a.txt', 'w', options={'x-goog-acl': 'project-private'}, retry_params='rp')
-        m.write.assert_called_with('1111111')
+    #
```
Comment (yaelmi3) [inline tests/unit/test_gcstorage.py:36]: why like this? we love tests :)

File: gaepypi/_decorators.py (hunk 4 above)
Comment (jefimm) [reply gaepypi/_decorators.py:35]: this is original syntax

File: tests/unit/test_gcstorage.py (hunk 14 above)
Comment (jefimm) [reply tests/unit/test_gcstorage.py:36]: this is old lib that is no longer relevant

**Review by yaelmi3 — APPROVED** [review]

**Pull request merged**